src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs

```diff
@@ -516,6 +524,113 @@ private async Task HandleChallengeAsyncInternal(AuthenticationProperties propert
+            await Events.PushAuthorization(context);
+
+            // If the event handled auth, we skip out default auth behavior
+            if (context.HandledClientAuthentication)
+            {
+                Logger.PushAuthorizationHandledClientAuthentication();
+            }
+            else
+            {
+                SetClientAuthenticationHeader(parRequest);
+            }
+
+            // If the event handled the push, there's nothing left to do and we bail out
+            if (context.HandledPush)
+            {
+                Logger.PushAuthorizationHandledPush();
+                return;
+            }
+
+            var parResponseMessage = await Backchannel.SendAsync(parRequest, Context.RequestAborted);
+
+            var requestUri = await GetRequestUri(parResponseMessage);
+            authorizeRequest.Parameters.Clear();
+            authorizeRequest.Parameters.Add("client_id", Options.ClientId);
+            authorizeRequest.Parameters.Add("request_uri", requestUri);
+        }
+        else
+        {
+            throw new InvalidOperationException("Attempt to push authorization with no pushed authorization endpoint configured.");
+        }
+    }
+
+    private async Task<string> GetRequestUri(HttpResponseMessage parResponseMessage)
+    {
+        // Check content type
+        var contentType = parResponseMessage.Content.Headers.ContentType;
+        if (!(contentType?.MediaType?.Equals("application/json", StringComparison.OrdinalIgnoreCase) ?? false))
+        {
+            throw new InvalidOperationException("Invalid response from pushed authorization: content type is not application/json.");
+        }
+
+        // Parse response
+        var parResponseString = await parResponseMessage.Content.ReadAsStringAsync();
+        var message = new OpenIdConnectMessage(parResponseString);
+
+        var requestUri = message.GetParameter("request_uri");
+        if(requestUri == null)
+        {
+            throw CreateOpenIdConnectProtocolException(message, parResponseMessage);
+        }
+        return requestUri;
+    }
+
+    private void SetClientAuthenticationHeader(HttpRequestMessage parRequest)
+    {
+        if (string.IsNullOrEmpty(Options.ClientId))
+        {
+            throw new InvalidOperationException("Missing client id");
+        }
+        if (string.IsNullOrEmpty(Options.ClientSecret))
+        {
+            throw new InvalidOperationException("Missing client secret");
+        }
+        var escapedClientId = Uri.EscapeDataString(Options.ClientId);
+        var escapedClientSecret = Uri.EscapeDataString(Options.ClientSecret);
+        var credential = $"{escapedClientId}:{escapedClientSecret}";
+        var encodedCredential = Convert.ToBase64String(Encoding.UTF8.GetBytes(credential));
+
+        parRequest.Headers.Authorization = new AuthenticationHeaderValue("basic", encodedCredential);
```

Comment:
Why was this header chosen for authentication by default? Looking at [RFC 9126](https://datatracker.ietf.org/doc/html/rfc9126#section-2), it allows authenticating the same way you would with the token endpoint as defined in [RFC 6749](https://datatracker.ietf.org/doc/html/rfc6749#section-3.2.1). 

I don't see anything where the client ID and client secret are put into a basic auth header. The "introductory example" in RFC 9126 uses "JWT client assertion-based authentication" defined in [RFC 7523](https://datatracker.ietf.org/doc/html/rfc7523#section-2.2), but that doesn't seem to provide a mechanism to acquire the JWT.

The common thing for the token endpoint seems to be to put the client ID and client secret in the request body of the token endpoint even though it's "NOT RECOMMENDED and SHOULD be limited to clients unable to directly utilize the HTTP Basic authentication scheme."

Why not just include the client secret in the `parRequest` body via `authorizeRequest.Parameters` since it's never seen by the client? It already includes the client ID, the spec seems to allow it, and it's effectively what we do by default for the token endpoint, right?

> The request also includes, as appropriate for the given client, any additional parameters necessary for client authentication (e.g., client_secret or client_assertion and client_assertion_type)

https://datatracker.ietf.org/doc/html/rfc9126#section-2.1

Reply:
I think you're right - I'll switch to the request body style.

I was trying to follow the "NOT RECOMMENDED" language, despite the inconsistency with the way we currently call the token endpoint. I had some thought that we might want to change _that_ as well, but the latest draft of OAuth 2.1 actually [drops ](https://author-tools.ietf.org/iddiff?url1=draft-ietf-oauth-v2-1-10&url2=draft-ietf-oauth-v2-1-11&difftype=--html) this language:

> Change client credentials to be required to be supported in the request body to avoid HTTP Basic authentication encoding interop issues

So that really makes it a moot point, and we should definitely just use the request body.

Reply:
Having basic auth of any description would be rather problematic from a security POV :)